karapace/compatibility/jsonschema/utils.py

```diff
@@ -86,11 +79,18 @@ def maybe_get_subschemas_and_type(schema: Any) -> Optional[Tuple[List[Any], Subs
 
         return (normalized_schemas, Subschema.ANY_OF)
 
-    for type_ in (Subschema.ALL_OF, Subschema.ANY_OF, Subschema.ONE_OF, Subschema.NOT):
+    for type_ in (Subschema.ALL_OF, Subschema.ANY_OF, Subschema.ONE_OF):
         subschema = schema.get(type_.value)
         if subschema is not None:
+            # https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.10.2.1
+            assert isinstance(subschema, list), "allOf/anyOf/oneOf must be an array"
```

Comment:
This was caught by mypy, not by a test.

I had to add the `isinstance' check to narrow the type, however that assert fails for `NOT`, so the two cases had to be split and it fixed the bug with `not` (I probably need to also add a unit test for that)